### ISSUE TITLE
CppOptimizer runStep ignores its learning rate parameter

### ISSUE THREAD

**Issue by RichardWarfield** [hasktorch/hasktorch#467]
I'm not sure how to fix this, or whether it should be fixed given the learning rate is provided in the <Optimizer>Options record.

```haskell
  runStep paramState optState lossValue lr = do
    performGC
    unsafeStep paramState optState lossValue
```

**Comment by junjihashimoto** [comment]
The learning rate is included in not optmizer's base class but child class.
https://github.com/pytorch/pytorch/blob/master/torch/csrc/api/include/torch/optim/optimizer.h
https://github.com/pytorch/pytorch/blob/master/torch/csrc/api/include/torch/optim/sgd.h#L26
So I think we should change step-function's definition or use upcast to update lr on unsafeStep.

**Comment by RichardWarfield** [comment]
I was thinking maybe just don't provide an `Optimizer` instance for CppOptimizerState and instead have a standalone `runStep` function (or have it be part of the CppOptimizer class)?  As it stands it seems like an awkward fit since two of the three class members of `Optimizer` are undefined.  Also I've been trying to write a typed version for `CppOptimizer` and this seems like an even more awkward fit since things like the `gradients` type don't show up anywhere.

**Comment by junjihashimoto** [comment]
Sorry for lack of explanation. 
There are 2 class, one is Haskell, another is C++.
Above my using class is c++'s one, and we may be able to add learning rate parameter for runStep function.

You know, Haskell's Optimizer's step and runStep' are not implemented.
The step function is not used except for runStep.

The gradient of `runStep'` can not be implemented, because c++'s step function can use loss only.
https://github.com/pytorch/pytorch/blob/master/torch/csrc/api/include/torch/optim/optimizer.h#L107

Haskell
```
class Optimizer optimizer where
    step :: LearningRate -> Gradients -> [Tensor] -> optimizer -> ([Tensor], optimizer)
    runStep :: (Parameterized model) => model -> optimizer -> Loss -> LearningRate -> IO (model, optimizer)
    runStep' :: (Parameterized model) =>model -> optimizer -> Gradients -> LearningRate -> IO (model, optimizer)

instance {-# OVERLAPS #-} CppOptimizer option => Optim.Optimizer (CppOptimizerState option) where
  step = error  "step is not implemented for CppOptimizer."
  runStep paramState optState lossValue lr = do
    performGC
    unsafeStep paramState optState lossValue
  runStep' = error  "runStep' is not implemented for CppOptimizer."
```

C++
https://github.com/pytorch/pytorch/blob/master/torch/csrc/api/include/torch/optim/optimizer.h

**Comment by junjihashimoto** [comment]
If we set sum (gradient * models' parameter) to loss, it is the same as using gradient's argument for runStep.
